bundles/org.openhab.binding.teleinfo/src/main/java/org/openhab/binding/teleinfo/internal/reader/io/TeleinfoInputStream.java

```diff
@@ -0,0 +1,673 @@
+        this(teleinfoInputStream, defaultTimeoutNextHeaderFrame, defaultTimeoutReadingFrame, false);
+    }
+
+    public TeleinfoInputStream(final InputStream teleinfoInputStream, boolean autoRepairInvalidADPSgroupLine) {
+        this(teleinfoInputStream, defaultTimeoutNextHeaderFrame, defaultTimeoutReadingFrame,
+                autoRepairInvalidADPSgroupLine);
+    }
+
+    public TeleinfoInputStream(final @Nullable InputStream teleinfoInputStream, long waitNextHeaderFrameTimeoutInMs,
+            long readingFrameTimeoutInMs, boolean autoRepairInvalidADPSgroupLine) {
+        if (teleinfoInputStream == null) {
+            throw new IllegalArgumentException("Teleinfo inputStream is null");
+        }
+
+        this.waitNextHeaderFrameTimeoutInMs = waitNextHeaderFrameTimeoutInMs;
+        this.readingFrameTimeoutInMs = readingFrameTimeoutInMs;
+        this.autoRepairInvalidADPSgroupLine = autoRepairInvalidADPSgroupLine;
+
+        this.bufferedReader = new BufferedReader(new InputStreamReader(teleinfoInputStream, StandardCharsets.US_ASCII));
+
+        groupLine = null;
+    }
+
+    @Override
+    public void close() throws IOException {
+        logger.debug("close() [start]");
+        bufferedReader.close();
+        executorService.shutdownNow();
+        super.close();
+        logger.debug("close() [end]");
+    }
+
+    /**
+     * Returns the next frame.
+     *
+     * @return the next frame or null if end of stream
+     * @throws InvalidFrameException if the read data from
+     * @throws TimeoutException if the delay to read a complete frame is expired (33,4 ms) or if the delay to find
+     *             the
+     *             header of next frame is expired (33,4 ms)
+     * @throws IOException
+     */
+    public synchronized @Nullable Frame readNextFrame() throws InvalidFrameException, TimeoutException, IOException {
+
+        // seek the next header frame
+        Future<@Nullable Void> seekNextHeaderFrameTask = executorService.submit(() -> {
+            while (!isHeaderFrame(groupLine)) {
+                groupLine = bufferedReader.readLine();
+                if (logger.isTraceEnabled()) {
+                    logger.trace("groupLine = {}", groupLine);
+                }
+                if (groupLine == null) { // end of stream
+                    logger.trace("end of stream reached !");
+                    return null;
+                }
+            }
+
+            logger.trace("header frame found !");
+            return null;
+        });
+        try {
+            logger.debug("seeking the next header frame...");
+            logger.trace("waitNextHeaderFrameTimeoutInMs = {}", waitNextHeaderFrameTimeoutInMs);
+            seekNextHeaderFrameTask.get(waitNextHeaderFrameTimeoutInMs, TimeUnit.MICROSECONDS);
+
+            if (groupLine == null) { // end of stream
+                return null;
+            }
+        } catch (InterruptedException e) {
+            throw new IllegalStateException(e);
```

Comment:
If I see correctly, you come here when `TeleinfoSerialControllerHandler.stopReceivingAndCloseSerialPort()` is invoked. The receiving thread needs to be interrupted, then, otherwise `join()` will block.